### PR TITLE
Use more generics and remove more code from mako files

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17186)

### CODE REVIEW

**Comment by nox** [comment]
r? @emilio 
@bors-servo try
https://treeherder.mozilla.org/#/jobs?repo=try&revision=e83db3a822bb35d19c3820cddeee398118a46f8a

File: components/style/values/generics/box.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Generic types for box properties.
+
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Copy, Debug, HasViewportPercentage, PartialEq, ToComputedValue, ToCss)]
+pub enum VerticalAlign<LengthOrPercentage>
```
Comment (emilio) [inline components/style/values/generics/box.rs]: Remove?

File: components/style/values/generics/gecko.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//! Generic types for legacy Gecko-only properties that should probably be
+//! unshipped at some point in the future.
+
+use std::fmt;
+use style_traits::ToCss;
+
+/// A generic value for scroll snap points.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Copy, Debug, HasViewportPercentage, PartialEq, ToComputedValue)]
+pub enum ScrollSnapPoint<LengthOrPercentage> {
```
Comment (emilio) [inline components/style/values/generics/gecko.rs]: This is in a gecko-only module, so no need for a `cfg_attr(feature = "servo"...)`. Though any reason why compiling it for both hurts?

**Comment by jdm** [comment]
```
   Compiling style v0.0.1 (file:///C:/buildbot/slave/windows-msvc-dev/build/components/style)
error[E0432]: unresolved import `values::specified::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16359:17
      |
16359 |         pub use values::specified::ScrollSnapPoint as SpecifiedValue;
      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::specified`

error[E0432]: unresolved import `values::computed::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16361:21
      |
16361 |             pub use values::computed::ScrollSnapPoint as T;
      |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::computed`

error[E0432]: unresolved import `values::specified::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16521:17
      |
16521 |         pub use values::specified::ScrollSnapPoint as SpecifiedValue;
      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::specified`

error[E0432]: unresolved import `values::computed::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16523:21
      |
16523 |             pub use values::computed::ScrollSnapPoint as T;
      |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `computed`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16363:69
      |
16363 |         #[inline] pub fn get_initial_value() -> computed_value::T { computed::ScrollSnapPoint::none() }
      |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `specified`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16369:13
      |
16369 |             specified::ScrollSnapPoint::parse(context, input)
      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `specified`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `computed`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16525:69
      |
16525 |         #[inline] pub fn get_initial_value() -> computed_value::T { computed::ScrollSnapPoint::none() }
      |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `specified`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16531:13
      |
16531 |             specified::ScrollSnapPoint::parse(context, input)
      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `specified`

error: type `properties::DeclaredValue<'_, _>` cannot be dereferenced
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16407:31
      |
16407 |                         match *value {
      |                               ^^^^^^

error: type `properties::DeclaredValue<'_, _>` cannot be dereferenced
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16569:31
      |
16569 |                         match *value {
      |                               ^^^^^^

error: aborting due to previous error(s)

error: Could not compile `style`.
```

File: components/style/values/specified/gecko.rs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+/// A specified type for scroll snap points.
+pub type ScrollSnapPoint = GenericScrollSnapPoint<LengthOrPercentage>;
+
+impl Parse for ScrollSnapPoint {
+    fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        if input.try(|i| i.expect_ident_matching("none")).is_ok() {
+            return Ok(GenericScrollSnapPoint::None);
+        }
+        input.expect_function_matching("repeat")?;
+        let length = input.parse_nested_block(|i|
```
Comment (emilio) [inline components/style/values/specified/gecko.rs]: nit: Wrap on braces?

**Review by emilio — APPROVED** [review]
r=me

File: components/style/values/specified/gecko.rs (hunk 3 above)
Comment (nox) [inline components/style/values/specified/gecko.rs]: Yes.

**Pull request merged**